doc/presentations/golangmtl-march-2016/presentation.md

```diff
@@ -0,0 +1,292 @@
+
+---
+
+# Performance
+
+- Large, dense `switch` is a binary search
+
+```go
+		switch i := ci.step(); i.opCode() {
+		case opMove:
+			frame[i.a()] = frame[i.b()]
+		case opLoadConstant:
+			frame[i.a()] = constants[i.bx()]
+		case opLoadConstantEx:
+			frame[i.a()] = constants[expectNext(ci, opExtraArg).ax()]
+		case opLoadBool:
+			frame[i.a()] = i.b() != 0
+			if i.c() != 0 {
+				ci.skip()
+			}
+		...
+		}
+```
+
+^ The core bytecode interpreter is a large dense switch statement, with a case for each opcode. C compilers typically turn this into a jump table - a table where each entry is the address of the code block for one case - and an indirect branch - that is, the value we're switching on is used to index the jump table in a branch instruction.
+
+---
+
+# Performance
+
+- Large, dense `switch` is a binary search
+
+```x86asm
+		SHRL $0x0, DX   ; extract opcode from instruction
+		ANDL $0x3f, DX
+		CMPQ $0x14, DX  ; op > 14?
+		JA 0xa2186
+		CMPQ $0x9, DX   ; op > 9?
+		JA 0x9ea87
+		CMPQ $0x4, DX   ; op > 4?
+		JA 0x9dd01
+		CMPQ $0x1, DX   ; op > 1?
+		JA 0x9da94
+		CMPQ $0x0, DX   ; op != 0?
+		JNE 0x9da09
+```
+
+^ Go implements the same thing as a binary search followed by a linear search when fewer than 4 cases remain. In go-lua, this means 2-5 compare & branch pairs for instruction dispatch.
+
+---
+
+# Performance
+
+- Dynamic allocation is slow for stack frames
+  - Control frames are a doubly-linked list of `callInfo` structs
+  - A misguided attempt to optimize memory footprint
+- Alternatives:
+  - Slice of `callInfo` containing both Go & Lua frame info
+  - Interleave control & data stack
+
+^ C Lua uses a C `union` to support native callout and Lua stack frames in the same `struct`
+
+^ Go doesn't provide `union` data types, so a series of refactors left us with dynamic allocation of both the common and variant pieces of a control stack frame. We cache stack frames, so if you're calling a series of Lua functions XOR a series of Go functions, then you don't pay for allocation, but the moment you switch from a Lua callee to a Go callee, for example, you blow away the cache.
+
+^ That said, we'd be much better off eating the memory overhead and storing control stack frames in a slice or interleaving control and data stacks.
+
+---
+
+# Performance
+
+> insert :dog: graphs here
+
+^ Genghis workers start over 300 goroutines per second on a c3.4xlarge instance, each with its own Lua VM that initializes, parses & executes the flow
+
+^ Bottleneck is usually network IO, not CPU and certainly not memory
+
+---
+
+# Current status
+
+- Actively used in `Shopify/genghis`
+- 28 forks, 2 ahead of `Shopify/go-lua`
+- Not actively developed right now
+  - `#patcheswelcome` :smile:
+- `Shopify/goluago` fills many gaps
```

Comment:
I'd add a slide just for goluago explaining what it adds and differences between the Lua API.